File: internal/agent/graph/state.go
package graph

import (
	"encoding/json"

	agentcontext "github.com/cybertortuga/aitriage/internal/agent/context"
	"github.com/cybertortuga/aitriage/internal/agent/llm"
	"github.com/cybertortuga/aitriage/internal/engine/core"
	"github.com/cybertortuga/aitriage/internal/report/healthcheck"
	"github.com/cybertortuga/aitriage/internal/scanner/deployaudit"
	"github.com/cybertortuga/aitriage/internal/scanner/entropy"
	"github.com/cybertortuga/aitriage/internal/scanner/external"
	"github.com/cybertortuga/aitriage/internal/scanner/network"
	"github.com/cybertortuga/aitriage/internal/scanner/nfr"
)

// AgentState represents the shared state flowing through the orchestrator.
type AgentState struct {
	ProjectPath string
	DeepScan    bool

	// Deterministic Go findings
	CoreFindings     []core.CheckResult
	ExternalFindings []external.UnifiedFinding
	NFRFindings      []nfr.NFRFinding
	DeployFindings   []deployaudit.DeployFinding
	NetworkFindings  []network.NetworkFinding

	SecurityScore int
	SecurityGrade string

	// HealthCheck holds the unified, multi-source Security Health Check result.
	// It is the authoritative security posture score, recomputed after AI
	// triage so that False Positives no longer penalise the repository.
	HealthCheck healthcheck.Result
	Policy      healthcheck.Policy

	// Repository context (gathered by gatherRepoContext)
	RepoContext *agentcontext.RepoContext

	// Data from scanners that was previously lost
	CriticalFiles []entropy.CriticalFile
	HistoryLeaks  []entropy.HistoryLeak
	Diagram       string

	// Map-Reduce state
	EnrichedFindings []EnrichedFinding
	Batches          [][]EnrichedFinding // Deprecated: runWorkers removed (June 2026)
	TriagedResults   []string            // Deprecated: runWorkers removed (June 2026)

	// SecureCoder-enhanced fields
	ThreatModel         *ThreatModel         // Structured threat model analysis
	FindingDispositions []FindingDisposition // TP/FP/NR classification per finding
	PoCResults          []PoCResult          // PoC verification results

	// LLM usage tracking (accumulated across all Chat calls)
	TotalUsage llm.Usage

	// Outputs
	ReportMarkdown  string // Full report (includes FP rationale) → artifact
	SummaryMarkdown string // Actionable summary (TP + NR only) → GHA Step Summary
	AIFixSpec       string
}

// EnrichedFinding is a unified representation of any finding with its source snippet attached.
type EnrichedFinding struct {
	ID        string `json:"id"`
	Type      string `json:"type"` // "core", "external", "nfr", "deploy", "network"
	Source    string `json:"source,omitempty"`
	Severity  string `json:"severity"`
	File      string `json:"file,omitempty"`
	Line      int    `json:"line,omitempty"`
	Message   string `json:"message"`
	Snippet   string `json:"code_snippet,omitempty"`
	ExtraData string `json:"extra_data,omitempty"`
	VulnID    string `json:"vuln_id,omitempty"` // CS-XXX-NNN format
}

// ── SecureCoder Threat Model Types ───────────────────────────────────────────

// ThreatModel holds the structured output from the LLM threat model step.
type ThreatModel struct {
	ComponentOverview  string       `json:"component_overview"`
	EntryPoints        []EntryPoint `json:"entry_points"`
	TrustBoundaries    TrustBounds  `json:"trust_boundaries"`
	SensitiveDataPaths []DataPath   `json:"sensitive_data_paths"`
	PrivilegedActions  []PrivAction `json:"privileged_actions"`
	PriorityAreas      []string     `json:"priority_areas"`
}

// EntryPoint describes a point where external data enters the system.
type EntryPoint struct {
	Endpoint   string `json:"endpoint"`
	Type       string `json:"type"`
	Trusted    bool   `json:"trusted"`
	Validation string `json:"validation"`
}

// TrustBounds describes authentication/authorization assumptions.
type TrustBounds struct {
	Authentication string `json:"authentication"`
	Authorization  string `json:"authorization"`
	ImplicitTrust  string `json:"implicit_trust"`
}

// DataPath describes a flow of sensitive data through the system.
type DataPath struct {
	DataType    string `json:"data_type"`
	Source      string `json:"source"`
	Destination string `json:"destination"`
	Protection  string `json:"protection"`
}

// PrivAction describes a privileged operation.
type PrivAction struct {
	Action   string `json:"action"`
	Location string `json:"location"`
	Guard    string `json:"guard"`
}

// FindingDisposition records the TP/FP/NR classification for a single finding.
type FindingDisposition struct {
	FindingIndex int    `json:"finding_index"`
	FindingID    string `json:"finding_id"`
	Disposition  string `json:"disposition"` // "True Positive", "False Positive", "Needs Manual Review"
	Rationale    string `json:"rationale"`
	// Confidence is the model's self-reported certainty: "high" | "medium" | "low".
	Confidence string `json:"confidence,omitempty"`
	// DispositionSource records how the disposition was produced for the audit
	// trail: "llm" | "cache" | "deterministic" | "nr-fallback".
	DispositionSource string `json:"disposition_source,omitempty"`
	// Fingerprint is the stable content hash used for dedup/caching.
	Fingerprint string `json:"fingerprint,omitempty"`
}

// PoCResult holds reasoning-based PoC verification for a finding.
type PoCResult struct {
	VulnerabilityType string    `json:"vulnerability_type"`
	Severity          string    `json:"severity"`
	AffectedFile      string    `json:"affected_file"`
	ReasoningSteps    []PoCStep `json:"reasoning_steps"`
	Conclusion        string    `json:"conclusion"`      // "Fix verified", "Fix incomplete", "Needs Manual Review"
	ExploitBlocked    *bool     `json:"exploit_blocked"` // nil = unknown
}

// PoCStep is one reasoning step in the PoC verification chain.
type PoCStep struct {
	// JSON numbers preserve labels such as 2.1 without weakening validation of
	// the surrounding PoC result.
	Step        json.Number `json:"step"`
	Description string      `json:"description"`
	Result      string      `json:"result"`
}

File: agent/state.py
"""
State definition for the AITriage security remediation agent.

Enhanced with SecureCoder-style workflow fields:
  - threat_model: structured threat analysis
  - finding_dispositions: True Positive / False Positive / Needs Review
  - security_plan: implementation plan with security verification section
  - poc_results: PoC verification results per finding
  - securecoder_findings: findings from SecureCoder scanner (when available)
  - findings_count_before: baseline count for fix_completed reporting
  - audit_report: CS-XXX-NNN formatted security audit report
"""
from __future__ import annotations

from typing import Annotated, Dict, List, Optional, TypedDict

from langgraph.graph.message import add_messages


class FindingDisposition(TypedDict, total=False):
    """Classification of a finding against the threat model."""

    finding: dict
    disposition: str  # "True Positive" | "False Positive" | "Needs Manual Review"
    rationale: str
    suppressed: bool  # True if auto-suppressed via SecureCoder API


class PoCResult(TypedDict, total=False):
    """PoC verification result for a single finding."""

    finding: dict
    vulnerability_type: str
    severity: str
    fix_summary: str
    reasoning_steps: List[dict]  # [{step, description, result}]
    conclusion: str  # "Fix verified" | "Fix incomplete"
    exploit_blocked: bool


class SecurityState(TypedDict):
    """
    TypedDict state container for the security remediation workflow.

    Lifecycle (SecureCoder-enhanced):
        scan → threat_model → analyze → plan → fix → verify → poc → report
    """

    project_path: str
    """Absolute path to the project being audited."""

    # ── Core scan data ────────────────────────────────────────────────
    scan_report: Optional[dict]
    """Raw JSON report from aitriage_scan MCP tool."""

    findings: List[dict]
    """Filtered list of security findings from AITriage scan."""

    securecoder_findings: List[dict]
    """Findings from SecureCoder scanner (when IDE is running)."""

    # ── Threat model (from SecureCoder determine_threat_model) ────────
    threat_model: Optional[dict]
    """Structured threat model: entry points, trust boundaries, etc."""

    finding_dispositions: List[Dict]
    """Per-finding classification: True Positive / False Positive / Needs Review."""

    # ── Planning & fixing ─────────────────────────────────────────────
    security_plan: Optional[str]
    """Implementation plan with security verification section."""

    current_finding: Optional[dict]
    """The specific finding currently being remediated."""

    generated_fix: Optional[str]
    """LLM-generated code patch for current_finding."""

    fix_verified: bool
    """True when the re-scan confirms the fix resolved the finding."""

    attempts: int
    """Number of fix attempts for current_finding (max 3 before give_up)."""

    # ── Verification & reporting ──────────────────────────────────────
    poc_results: List[Dict]
    """PoC verification results per finding (reasoning, not execution)."""

    findings_count_before: int
    """Baseline finding count (for SecureCoder /fix_completed reporting)."""

    audit_report: Optional[str]
    """CS-XXX-NNN formatted security audit report."""

    # ── LLM state ─────────────────────────────────────────────────────
    messages: Annotated[list, add_messages]
    """Chat history with the LLM (used for multi-turn reasoning)."""

    final_report: Optional[str]
    """Human-readable markdown summary of all actions taken."""
